### PR TITLE
CLICKHOUSE-1420 Clear query in client by Ctrl+C

### PR DESCRIPTION
**Pull request by CurtizJ**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

File: dbms/programs/client/Client.cpp [hunk 1]
```diff
@@ -658,6 +659,24 @@ class Client : public Poco::Util::Application
         return boost::replace_all_copy(prompt_by_server_display_name, "{database}", config().getString("database", "default"));
     }
 
+    static void signalHandler(int signo)
+    {
+        if(signo == SIGINT)
+        {
+            if (strcmp(rl_line_buffer, "") == 0)
```
Comment (alexey-milovidov) [inline dbms/programs/client/Client.cpp]: check for first byte is enough

File: dbms/programs/client/Client.cpp [hunk 2]
```diff
@@ -658,6 +659,24 @@ class Client : public Poco::Util::Application
         return boost::replace_all_copy(prompt_by_server_display_name, "{database}", config().getString("database", "default"));
     }
 
+    static void signalHandler(int signo)
+    {
+        if(signo == SIGINT)
```
Comment (alexey-milovidov) [inline dbms/programs/client/Client.cpp]: Style

File: dbms/programs/client/Client.cpp [hunk 3]
```diff
@@ -658,6 +659,24 @@ class Client : public Poco::Util::Application
         return boost::replace_all_copy(prompt_by_server_display_name, "{database}", config().getString("database", "default"));
     }
 
+    static void signalHandler(int signo)
```
Comment (alexey-milovidov) [inline dbms/programs/client/Client.cpp]: Missing comment.

File: dbms/programs/client/Client.cpp [hunk 4]
```diff
@@ -571,7 +571,8 @@ class Client : public Poco::Util::Application
                     Poco::File(history_file).createFile();
             }
 
-            loop();
+            signal(SIGINT, signalHandler);
```
Comment (alexey-milovidov) [inline dbms/programs/client/Client.cpp]: Error is not checked.

File: dbms/programs/client/Client.cpp [hunk 5]
```diff
@@ -658,6 +659,24 @@ class Client : public Poco::Util::Application
...
     }
 
+    static void signalHandler(int signo)
+    {
+        if(signo == SIGINT)
+        {
+            if (strcmp(rl_line_buffer, "") == 0)
+            {
+                std::cout << std::endl;
+                exit(SIGINT);
```
Comment (alexey-milovidov) [inline dbms/programs/client/Client.cpp]: Goodbye message is not printed.

File: dbms/programs/client/Client.cpp [hunk 6]
```diff
@@ -658,6 +659,24 @@ class Client : public Poco::Util::Application
...
+            }
+            else       
+            {   
+                std::cout << std::endl;
+                rl_replace_line("", 0);
+                rl_forced_update_display();
```
Comment (alexey-milovidov) [inline dbms/programs/client/Client.cpp]: Return value is ignored.

File: dbms/programs/client/Client.cpp [hunk 7]
```diff
@@ -1099,6 +1118,8 @@ class Client : public Poco::Util::Application
     /// Also checks if query execution should be cancelled.
     void receiveResult()
     {
+        signal(SIGINT, SIG_DFL);
```
Comment (alexey-milovidov) [inline dbms/programs/client/Client.cpp]: Is it really needed?
(We block signals in InterruptListener.)

Comment (CurtizJ) [reply dbms/programs/client/Client.cpp]: Without it exit by double press Ctrl+c doesn`t work.

File: dbms/programs/client/Client.cpp [hunk 8]
```diff
@@ -571,7 +571,9 @@ class Client : public Poco::Util::Application
                     Poco::File(history_file).createFile();
             }
 
-            loop();
+            if (signal(SIGINT, signalHandler) == SIG_ERR)
+                throwFromErrno("Cannot handle signal.", ErrorCodes::CANNOT_WAIT_FOR_SIGNAL);
```
Comment (alexey-milovidov) [inline dbms/programs/client/Client.cpp]: Wrong error code.

File: dbms/programs/client/Client.cpp [hunk 9]
```diff
@@ -658,6 +660,26 @@ class Client : public Poco::Util::Application
...
 
+    /// Handle interrupt signal to clear current query by pressing Ctrl+C
+    /// or to exit if query is empty
+    static void signalHandler(int signo)
+    {
+        if (signo == SIGINT)
+        {
+            if (!rl_line_buffer[0])
+            {
+                std::cout << std::endl << "Bye." << std::endl;
```
Comment (alexey-milovidov) [inline dbms/programs/client/Client.cpp]: The code is duplicated.

File: dbms/programs/client/Client.cpp [hunk 10]
```diff
@@ -658,6 +660,26 @@ class Client : public Poco::Util::Application
...
+                exit(SIGINT);
```
Comment (alexey-milovidov) [inline dbms/programs/client/Client.cpp]: Destructor if Client class wont be invoked. ASan and Valgrind will report false positive memory leaks.

File: dbms/programs/client/Client.cpp (hunk 7 above)
Comment (alexey-milovidov) [reply dbms/programs/client/Client.cpp]: How it can be explained?

File: dbms/programs/client/Client.cpp [hunk 11]
```diff
@@ -1129,6 +1153,8 @@ class Client : public Poco::Util::Application
 
         if (cancelled && is_interactive)
             std::cout << "Query was cancelled." << std::endl;
+            
+        signal(SIGINT, signalHandler);
```
Comment (alexey-milovidov) [inline dbms/programs/client/Client.cpp]: Return code is not checked.

File: dbms/programs/client/Client.cpp [hunk 12]
```diff
@@ -658,6 +660,26 @@ class Client : public Poco::Util::Application
         return boost::replace_all_copy(prompt_by_server_display_name, "{database}", config().getString("database", "default"));
     }
 
+    /// Handle interrupt signal to clear current query by pressing Ctrl+C
+    /// or to exit if query is empty
+    static void signalHandler(int signo)
+    {
+        if (signo == SIGINT)
```
Comment (alexey-milovidov) [inline dbms/programs/client/Client.cpp]: Useless check.

File: dbms/programs/client/Client.cpp (hunk 9 above)
Comment (alexey-milovidov) [inline dbms/programs/client/Client.cpp]: It is not signal-safe.

**Pull request closed without merging**